File: trie.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>
#include <queue>

typedef struct TrieNode
{
    std::string word; // word in the node.
    std::unordered_map<char, TrieNode *> children;
    TrieNode(const std::string &w = "") : word(w) {}
} TrieNode;

TrieNode *Insert(TrieNode *root, const std::string &word)
{
    if (root == nullptr)
    {
        root = new TrieNode;
    }

    TrieNode *current = root;
    for (char c : word)
    {
        if (current->children.count(c))
        {
            current = current->children[c];
        }
        else
        {
            current->children[c] = new TrieNode;
            current = current->children[c];
        }
    }

    // todo: where to put the word in a node
    if (!word.empty())
    {
        current->word = word;
    }
    return root;
}

bool Contains(TrieNode *root, const std::string &word)
{
    TrieNode *current = root;
    for (const char &c : word)
    {
        if (current->children.count(c))
        {
            current = current->children[c];
        }
    }
    if (!word.empty() && current->word == word)
    {
        return true;
    }
    return false;
}

std::vector<std::string> AutoCompleteSuggestions(TrieNode *root, const std::string &partial)
{
    std::vector<std::string> suggestions;
    TrieNode *current = root;
    for (const char &c : partial)
    {
        if (current->children.count(c))
        {
            if (current->word == partial)
            {
                suggestions.push_back(current->word);
                return suggestions;
            }

            current = current->children[c];
        }
    }

    // the partial is finished.. now perform BFS (level order traversal)
    std::queue<TrieNode *> q;
    q.push(current);
    while (!q.empty())
    {
        TrieNode *t = q.front();
        q.pop();

        if (!t->word.empty())
        {
            suggestions.push_back(t->word);
        }

        for (const auto &it : t->children)
        {
            q.push(it.second);
        }
    }

    return suggestions;
}

TrieNode *DeleteWord(TrieNode *root, const std::string &word)
{
    TrieNode **prev = nullptr;
    TrieNode *current = root;
    for (const char &c : word)
    {
        if (current->children.count(c))
        {
            prev = &current;
            current = current->children[c];
        }
    }

    if (!word.empty() && current != nullptr)
    {
        current->word = "";
        if (current->children.size() == 0)
        {
            *prev = nullptr;
            delete current;
        }
    }
    return root;
}

int main()
{
    std::vector<std::string> dict{"peter", "piper", "picked", "peck", "pickled", "peppers"};
    TrieNode *root = new TrieNode;

    for (const auto &word : dict)
    {
        root = Insert(root, word);
    }

    std::string word = "pickled";
    std::cout << "Looking for word " << word << " " << Contains(root, word) << "\n";

    std::cout << "Auto complete suggestions \n";
    auto suggestions = AutoCompleteSuggestions(root, "pic");
    for (const auto w : suggestions)
    {
        std::cout << w << " ";
    }

    root = DeleteWord(root, word);
    std::cout << "\nAfter deleting word, lookup " << word << " " << Contains(root, word) << "\n";
}

File: golang/datastructures/trie/trie.go
package trie

import "container/list"

type node struct {
	word     string
	children map[rune]*node
}

// Trie ... Trie type
type Trie struct {
	root *node
}

func newNode() *node {
	return &node{word: string(""), children: make(map[rune]*node)}
}

// New ... Creates a new instance of a Trie
func New() *Trie {
	n := newNode()
	return &Trie{n}
}

// Insert ... Insert a word in a trie
func (t *Trie) Insert(word string) {
	current := t.root
	for _, c := range word {
		if _, ok := current.children[c]; ok {
			current = current.children[c]
		} else {
			current.children[c] = newNode()
			current = current.children[c]
		}
	}

	if len(word) != 0 {
		current.word = word
	}
}

// Contains ... Contains a specific word
func (t *Trie) Contains(word string) bool {
	current := t.root
	for _, w := range word {
		if _, ok := current.children[w]; ok {
			current = current.children[w]
		}
	}

	if len(word) != 0 && current.word == word {
		return true
	}
	return false
}

// Suggestions ... Suggest auto-completing words
func (t *Trie) Suggestions(partial string) []string {
	var suggestions []string
	current := t.root
	for _, c := range partial {
		if _, ok := current.children[c]; ok {
			if current.word == partial {
				suggestions = append(suggestions, current.word)
				return suggestions
			}
			current = current.children[c]
		}
	}

	// we have searched for every word in a partial string.
	// Now perform BFS to compute all the suggestions.

	q := list.New() // as a queue
	q.PushBack(current)
	for q.Len() != 0 {
		e := q.Front()
		q.Remove(e)

		t := e.Value.(*node)
		if len(t.word) != 0 {
			suggestions = append(suggestions, t.word)
		}

		for _, v := range t.children {
			q.PushBack(v)
		}
	}
	return suggestions
}

// DeleteWord ... Delete a word from the trie
func (t *Trie) DeleteWord(word string) {
	if len(word) == 0 {
		return
	}

	current, prev := t.root, &(t.root)
	for _, c := range word {
		if _, ok := current.children[c]; ok {
			prev = &current
			current = current.children[c]
		}
	}

	if current != nil {
		current.word = ""
		if len(current.children) == 0 {
			prev, current = nil, nil
		}
	}
	_ = prev
}
